File: kundol/code_by_jvm/app/src/main/kotlin/io/philo/codingtest/ch_1__big_o/_1__이론/1 누적합.kt
package io.philo.codingtest.ch_1__big_o._1__이론

class _1__누적합

//val arr: Array<IntRange> = arrayOf(1..10)
val arr: Array<Int> = arrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

fun main() {

    val result = sub_sum(1, 4)
    println("result = ${result}")
}

fun sub_sum(st: Int, ed: Int): Int {

    val sum_arr: Array<Int> = Array(11) { 0 }

    sum_arr[1] = arr[1]

    for (i in 2..ed) {
        sum_arr[i] += sum_arr[i-1] + arr[i]
    }

    return sum_arr[ed] - sum_arr[st]
}

File: kundol/code_by_python/pythonProject/kundol/ch_1__Big_O/_1__문제/1 누적합.py
def sub_sum(arr, st, ed):
    arr_size = len(arr)
    sum_arr = [0] * (arr_size + 1)

    # s0=a0, s1=s0+a1, s2=s1+a2, s3=s2+a3
    for i in range(0, arr_size-1):
        sum_arr[i+1] = sum_arr[i] + arr[i+1]

    return sum_arr[ed] - sum_arr[st]

# 1 2 3 4 5
# 3~1
# 2+3+4
# 1~4 - 1~1 = s3 - s1
if __name__ == "__main__":
    arr = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    result = sub_sum(arr, 1, 4)
    print("result =", result)
